net/pubsub.go

```diff
@@ -0,0 +1,165 @@
+package net
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/libp2p/go-libp2p-core/peer"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/textileio/go-threads/core/thread"
+	pb "github.com/textileio/go-threads/net/pb"
+)
+
+// TopicHandler receives all pushed thread records.
+type TopicHandler func(*pb.PushRecordRequest)
+
+// TopicManager manages thread pubsub topics.
+type TopicManager struct {
+	sync.RWMutex
+
+	ctx    context.Context
+	host   peer.ID
+	ps     *pubsub.PubSub
+	handle TopicHandler
+	m      map[thread.ID]*topic
+}
+
+type topic struct {
+	t *pubsub.Topic
+	h *pubsub.TopicEventHandler
+	s *pubsub.Subscription
+}
+
+// NewTopicManager returns a new thread topic manager.
+func NewTopicManager(ctx context.Context, host peer.ID, ps *pubsub.PubSub, handler TopicHandler) *TopicManager {
+	return &TopicManager{
+		ctx:    ctx,
+		host:   host,
+		ps:     ps,
+		handle: handler,
```

Comment:
This is a callback used by the server, which in the end only cares to receive the single push record requests, which are passed onto the normal `PushRecord` method.